### PR TITLE
[remove nixpkgs] part 2: generate flake.nix using builtins.fetchClosure

### PR DESCRIPTION
**Pull request by savil**
## Summary

See #1208 for motivation.
That previous PR handled the "writing the system info to the lockfile". In this PR: 

1. Generate the `flake.nix` from `flake_remove_nixpkgs.nix.tmpl` template.
   - this new template uses `builtins.fetchClosure` to download the package's binary directly from the binary-cache-store (like `cache.nixos.org`). This lets us bypass `nixpkgs` for nix packages in the binary store.
2. The template data is `shellgen.flakePlan`.
3. In the `flakePlan`, we:
   a.  filter `flakeInputs` to be those _not_ in the binary store (like remote or local flakes)
   b. add a new field `Packages []*nix.Package`.
   
In `nix.Package`, we have a few new methods specific to the BinaryStore, like `IsInBinaryStore` and `PathInBinaryStore`.

**Testing note:**
I was stymied in extensively testing this by:
- search index lacking x86_64-darwin system info (and I have an intel mac)
- devbox cloud having limitations (will share details elsewhere)
- the simple test plan below took me many hours to complete.

cc @mikeland73 

## How was it tested?

Tested with `examples/development/go/hello-world`.

For a more complex example, with a flake, the `examples/flakes/remote` produces this `flake.nix`:
```
{
   description = "A devbox shell";

   inputs = {
     nixpkgs.url = "http://lax.devbox-nixed.internal/nixos/nixpkgs/archive/f80ac848e3d6f0c12c52758c0f25c10c97ca3b62.tar.gz";
     gh-nixos-nixpkgs-5233fd2ba76a3accb5aaa999c00509a11fd0793c.url = "github:nixos/nixpkgs/5233fd2ba76a3accb5aaa999c00509a11fd0793c";
     gh-F1bonacc1-process-compose.url = "github:F1bonacc1/process-compose";
   };

   outputs = {
     self,
     nixpkgs,
     gh-nixos-nixpkgs-5233fd2ba76a3accb5aaa999c00509a11fd0793c,
     gh-F1bonacc1-process-compose,
   }:
       let
          pkgs = nixpkgs.legacyPackages.x86_64-linux;
       in
       {
         devShells.x86_64-linux.default = pkgs.mkShell {
           buildInputs = [
             gh-nixos-nixpkgs-5233fd2ba76a3accb5aaa999c00509a11fd0793c.legacyPackages.x86_64-linux.hello
             gh-nixos-nixpkgs-5233fd2ba76a3accb5aaa999c00509a11fd0793c.legacyPackages.x86_64-linux.cowsay
             gh-F1bonacc1-process-compose.packages.x86_64-linux.process-compose
           ];
         };
       };
 }
```

### CODE REVIEW

**Comment by savil** [comment]
Current dependencies on/for this PR:
* main
  * **PR #1208** <a href="https://app.graphite.dev/github/pr/jetpack-io/devbox/1208" target="_blank">Graphite</a> 
    * **PR #1209** <a href="https://app.graphite.dev/github/pr/jetpack-io/devbox/1209" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/jetpack-io/devbox/1209?utm_source=stack-comment).

**Comment by gcurtis** [comment]
Sorry for the delay - I just remembered that I never reviewed this. I'll take a look tomorrow morning.

File: internal/shellgen/flake_package.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	if sysInfo == nil {
+		return nil, nil
+	}
+
+	attributePath, err := pkg.PackageAttributePath()
+	if err != nil {
+		return nil, err
+	}
+
+	storeDir := strings.Join([]string{sysInfo.FromHash, sysInfo.StoreName, sysInfo.StoreVersion}, "-")
```
Comment (gcurtis) [inline internal/shellgen/flake_package.go:32]: The store version might be empty, in which case omit the trailing -.

File: internal/shellgen/flake_package.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+	return &flakePackage{
+		Name:      attributePath,
+		FromStore: nixosCacheURL,
+		FromPath:  filepath.Join("/nix/store", storeDir),
+		// TODO add ToPath:
+	}, nil
+}
+
+func flakePackages(devbox devboxer) ([]*flakePackage, error) {
```
Comment (gcurtis) [inline internal/shellgen/flake_package.go:41]: I know this wasn't added in your PR, but the number of interfaces we have is making the code really hard to follow.

File: internal/shellgen/tmpl/flake_remove_nixpkgs.nix.tmpl [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+     {{- end }}
+   }:
+       let
+          pkgs = nixpkgs.legacyPackages.{{ .System }};
+          {{- range $_, $pkg := .Packages }}
+          {{ $pkg.Name }}.{{ $.System }} = builtins.fetchClosure{
+             fromStore = "{{ $pkg.FromStore }}";
+             fromPath = "{{ $pkg.FromPath }}";
+          };
+          {{- end }}
```
Comment (gcurtis) [inline internal/shellgen/tmpl/flake_remove_nixpkgs.nix.tmpl:26]: I think the attribute path order should be flipped to `{{ $pkg.System }}.{{ $.Name }}` to follow the flake convention. The name should also be quoted in case it has special characters.

Another option is to just inline this directly in the `builtInputs` list with something like:

```nix
[
   {{- range .Packages }}
   (builtins.fetchClosure{
      fromStore = "{{ .FromStore }}";
      fromPath = "{{ .FromPath }}";
   })
   {{- end }}
]
```

File: internal/shellgen/tmpl/flake_remove_nixpkgs.nix.tmpl [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+         devShells.{{ .System }}.default = pkgs.mkShell {
+           buildInputs = [
+             {{- range $_, $pkg := .Packages }}
+             {{ $pkg.Name }}.{{ $.System }}
+             {{- end }}
+             {{- range $_, $flake := .FlakeInputs }}
+             {{- range $flake.BuildInputs }}
+             {{.}}
+             {{- end }}
+             {{- end }}
```
Comment (gcurtis) [inline internal/shellgen/tmpl/flake_remove_nixpkgs.nix.tmpl:38]: Can be simplified to:

```suggestion
             {{- range .FlakeInputs }}
             {{- range .BuildInputs }}
             {{.}}
             {{- end }}
             {{- end }}
```

File: internal/shellgen/flake_plan.go [hunk 5]
```diff
@@ -3,13 +3,17 @@ package shellgen
...
+
+	"go.jetpack.io/devbox/internal/nix"
 )
 
 // flakePlan contains the data to populate the top level flake.nix file
 // that builds the devbox environment
 type flakePlan struct {
 	NixpkgsInfo *NixpkgsInfo
 	FlakeInputs []*flakeInput
+	Packages    []*flakePackage
```
Comment (gcurtis) [inline internal/shellgen/flake_plan.go:15]: Could this be a slice of values instead? It doesn't look like they need to be pointers.

File: internal/shellgen/flake_package.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+import (
+	"path/filepath"
+	"strings"
+
+	"go.jetpack.io/devbox/internal/boxcli/featureflag"
+	"go.jetpack.io/devbox/internal/nix"
+)
+
+type flakePackage struct {
```
Comment (gcurtis) [inline internal/shellgen/flake_package.go:11]: I'm not sure if `flakePackage` is an accurate name. It's fetching a store path and its dependencies, which may or may not have been built from a flake.

File: internal/shellgen/flake_package.go [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+	if err != nil {
+		return nil, err
+	}
+
+	storeDir := strings.Join([]string{sysInfo.FromHash, sysInfo.StoreName, sysInfo.StoreVersion}, "-")
+	return &flakePackage{
+		Name:      attributePath,
+		FromStore: nixosCacheURL,
+		FromPath:  filepath.Join("/nix/store", storeDir),
+		// TODO add ToPath:
```
Comment (gcurtis) [inline internal/shellgen/flake_package.go:37]: Have you looked at the [`builtins.storePath` function](https://nixos.org/manual/nix/stable/language/builtins.html#builtins-storePath)? If we're running in impure mode, then it might make things a little easier. It doesn't require a content-addressed hash and it will automatically use the Nix-configured binary caches (aka substituters). That'll make this work automatically with nixed in Devbox Cloud.

**Review by gcurtis — APPROVED** [review]
I had some comments, but approving to keep the train moving.

File: internal/shellgen/flake_input.go [hunk 8]
```diff
@@ -77,11 +78,18 @@ func flakeInputs(ctx context.Context, devbox devboxer) ([]*flakeInput, error) {
...
-	order := []string{}
 	// We prioritize plugin packages so that the php plugin works. Not sure
 	// if this is behavior we want for user plugins. We may need to add an optional
 	// priority field to the config.
-	for _, input := range append(pluginInputs, userInputs...) {
+	allInputs := append(pluginInputs, userInputs...)
+	allInputs = lo.Filter(allInputs, func(item *nix.Package, _ int) bool {
+		// This is temporary, so that we can support the existing flake.nix.tmpl
+		// as well as the new flake_remove_nixpkgs.nix.tmpl.
+		return !featureflag.RemoveNixpkgs.Enabled() || item.SystemInfo() == nil
```
[the remark this answers is not in the document]
Comment (savil) [reply internal/shellgen/flake_input.go:88]: This will continue to include packages (like local or remote flakes) that don't have system info

File: internal/shellgen/flake_package.go (hunk 6 above)
Comment (mikeland73) [reply internal/shellgen/flake_package.go:11]: I'm a bit confused, why do we need this instead of using `nix.Package` directly? It feels like just a another struct with the same data and function?

File: internal/shellgen/flake_plan.go (hunk 5 above)
Comment (mikeland73) [reply internal/shellgen/flake_plan.go:15]: Can it be `nix.Packages` instead?

File: internal/shellgen/tmpl/flake_remove_nixpkgs.nix.tmpl [hunk 9]
```diff
@@ -0,0 +1,42 @@
+{
+   description = "A devbox shell";
+
+   inputs = {
+     {{/* nixpkgs could be removed once we turn php/haskell planners into flakes */ -}}
```
Comment (mikeland73) [inline internal/shellgen/tmpl/flake_remove_nixpkgs.nix.tmpl:5]: rm this comment.

**Review by mikeland73 — COMMENTED** [review]
Not blocking, but I would really want to avoid adding a new `Package` struct. I also think system info should be part of resolve. Ideally system info can be abstracted away so that the whatever generates the flake can just get the data if it's present.

btw, you should assume that we will not be able to avoid nixpkgs for certain packages for the short and medium term, so we need to structure this in a clean way where both types can coexist.

**Comment by savil** [comment]
@gcurtis addressed comments and made updates

@mikeland73 Folded `flakePackages` into `nix.Package`. Addressed the concern around calling `lockfile.Resolve` in the getter for `FromPath` in `nix.Package`.

File: internal/nix/input.go [hunk 10]
```diff
@@ -73,7 +73,15 @@ func PackageFromString(raw string, locker lock.Locker) *Package {
...
-	return &Package{*pkgURL, locker, raw, ""}
+
+	pkg := &Package{
+		URL:                                 *pkgURL,
+		lockfile:                            locker,
+		Raw:                                 raw,
+		normalizedPackageAttributePathCache: "",
+	}
+
+	return pkg
```
Comment (mikeland73) [inline internal/nix/input.go:84]: nit, you can revert this if you want. You can also omit `normalizedPackageAttributePathCache` if you used name fields.

File: internal/nix/input.go [hunk 11]
```diff
@@ -407,6 +415,40 @@ func (p *Package) hashFromNixPkgsURL() string {
 	return HashFromNixPkgsURL(p.URLForFlakeInput())
 }
 
+// BinaryCacheStore is the store from which to fetch this package's binaries.
+// It is used as FromStore in builtins.fetchClosure.
+const BinaryCacheStore = "https://cache.nixos.org"
+
+func (p *Package) IsInBinaryStore() bool {
+	return p.isVersioned()
+}
```
Comment (mikeland73) [inline internal/nix/input.go:424]: Is there a better way to determine this? For example only true if lock file data has system info?

File: internal/shellgen/flake_plan.go [hunk 12]
```diff
@@ -35,9 +40,17 @@ func newFlakePlan(ctx context.Context, devbox devboxer) (*flakePlan, error) {
 		}
 	}
 
-	shellPlan := &flakePlan{}
-	var err error
-	shellPlan.FlakeInputs, err = flakeInputs(ctx, devbox)
+	userPackages := devbox.PackagesAsInputs()
+	pluginPackages, err := devbox.PluginManager().PluginInputs(userPackages)
```
Comment (mikeland73) [inline internal/shellgen/flake_plan.go:44]: I'll rename `PluginInputs` and `PackagesAsInputs` in a follow up.

Also, I'm thinking of changing plugins to encapsulate it more. Having plugins show up in shellgen package is not idael

**Review by mikeland73 — APPROVED** [review]
This is great! I really like how this ended up.

File: internal/nix/input.go (hunk 11 above)
Comment (savil) [reply internal/nix/input.go:424]: 🤔 yeah, we could add that check too. In my previous iteraiton of this PR, I was checking both `p.isVersioned()` and "has system info". 

I can add that here. It feels weird to possibly make network-calls in an if-statement but I understand the architecture better now so lets roll with it.
